bot.py

```diff
@@ -0,0 +1,191 @@
+                return await ctx.send(f'{bot_name} is being used in {channel}')
+            await channel.connect()
+
+    @commands.command()
+    async def play(self, ctx: commands.Context, *query: str) -> None:
+        # Handle either youtube link or query
+        if len(query) > 1:
+            query = " ".join(query)
+            # When its not given a link, it returns all possible links when this query is searched on yt
+            info_dict = ytdl.extract_info(query, download=False)
+            query = info_dict['entries'][0]['webpage_url']
+        else:
+            query = query[0]
+
+        if await self.is_inCall(ctx):
+            async with ctx.typing():
+                try:
+                    # checks if file exists in cache before calling yt extract
+                    youtube_info = self.try_to_retrieve_info(query)
+                    if youtube_info == None:
+                        youtube_info = ytdl.extract_info(
+                            url=query, download=False)
+                        self.cache_file(query, youtube_info)
+
+                    if not ctx.voice_client.is_playing():
+                        ctx.voice_client.play(FFmpegPCMAudio(
+                            youtube_info['url'], **ffmpeg_options), after=lambda e: self.bot.loop.create_task(self.play_next(ctx)))
+                        await ctx.send(f"Now playing {youtube_info['title']}")
+                    else:
+                        self.audio_queue.append(youtube_info)
+                        await ctx.send(f"{youtube_info['title']} queued")
+                except:
+                    await ctx.send("Invalid Source")
+
+    async def play_next(self, ctx: commands.Context) -> None:
+        # check if there is anything to play
+        if self.audio_queue:
+            audio_source = FFmpegPCMAudio(
+                (self.audio_queue[0]['url']), **ffmpeg_options)
+            await ctx.send(f"Now playing {self.audio_queue[0]['title']}")
+            self.audio_queue.popleft()
+            ctx.voice_client.play(
+                audio_source, after=lambda e: self.bot.loop.create_task(self.play_next(ctx)))
+
+    '''
+    Looks to see if yt extract info is in cache and pulls the dict from there instead of 
+    extracting from youtube link or searching the word query to get dict info again. 
+    '''
+
+    def try_to_retrieve_info(self, audioTitle: str) -> (dict | None):
+        if redis_mgr.exists(audioTitle):
+            yt_info = redis_mgr.hgetall(audioTitle)
+            yt_info = {field.decode(): yt_info[field].decode()
+                       for field in yt_info}
+            return yt_info
+
+    '''
+    Once a song is requested and if it isn't returned by the try_to_retrieve func then 
+    this function is called to cache the most recently request song.  
+    All we ever use is the title and url so just store those  
+    '''
+
+    def cache_file(self, query: str, ytExtract: dict) -> None:
```

Comment:
you could also try storing the audio, https://stackoverflow.com/questions/54040348/redis-hash-how-to-save-a-sound-file. It might be cheaper to get storage than network calls in whatever cloud provider you use. You could make it a flag on the app startup, to cache the soundbytes or not.